### PR TITLE
feat(card-impl): Implement card 'Inner Fire' (CS1_129)

### PR DESCRIPTION
**Pull request by 95D**
feat(card-impl) : Expert1Card "Inner Fire" ( CS1_129 , CS1_129e ) { Gen / Test }
feat(Enchants) : Shortcut Enchant SetAttackScriptTag, SetHealthScriptTag, AddAttackScriptTag
( * i did reference AddHealthScriptTag )
( * i think about that need to implement enchant for conditional set attack/ health .. like "Inner Fire" enchant )
fix : CodeFactor , VSBuild Warning ( no reference local value )
._.)

### CODE REVIEW

**Comment by 95D** [comment]
what is next my work ? could i implement more card?
._.)a

**Comment by utilForever** [comment]
Sorry. I'll review it tonight.

**Comment by 95D** [comment]
Please do it when you feel comfortable!
._.)b

File: Includes/Rosetta/Enchants/Enchants.hpp [hunk 1]
```diff
@@ -21,9 +21,19 @@ namespace RosettaStone
...
+    //! Enchant that adds attack and uses script tag.
+    inline static Enchant AddAttackScriptTag = 
+        Enchant(Effects::AttackN(0), true);
     //! Enchant that adds health and uses script tag.
     inline static Enchant AddHealthScriptTag =
         Enchant(Effects::HealthN(0), true);
+    
+    //! Enchant that set attack and uses script tag.
+    inline static Enchant SetAttackScriptTag =
+        Enchant(Effect(GameTag::ATK, EffectOperator::SET, 0),true);
```
Comment (utilForever) [inline Includes/Rosetta/Enchants/Enchants.hpp]: Check whitespace. Should be `0), true);`

File: Sources/Rosetta/CardSets/Expert1CardsGen.cpp [hunk 2]
```diff
@@ -4,12 +4,22 @@
...
+#include <Rosetta/Enchants/Enchants.hpp>
 #include <Rosetta/Tasks/SimpleTasks/DamageTask.hpp>
 #include <Rosetta/Tasks/SimpleTasks/DrawTask.hpp>
 #include <Rosetta/Tasks/SimpleTasks/HealTask.hpp>
 #include <Rosetta/Tasks/SimpleTasks/SetGameTagTask.hpp>
+#include <Rosetta/Tasks/SimpleTasks/FuncNumberTask.hpp>
+#include <Rosetta/Tasks/SimpleTasks/GetGameTagTask.hpp>
+#include <Rosetta/Tasks/SimpleTasks/MathSubTask.hpp>
+#include <Rosetta/Tasks/SimpleTasks/AddEnchantmentTask.hpp>
 #include <Rosetta/Tasks/SimpleTasks/SummonTask.hpp>
```
Comment (utilForever) [inline Sources/Rosetta/CardSets/Expert1CardsGen.cpp]: Please reorder header files in alphabetical order.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp [hunk 3]
```diff
@@ -198,4 +198,41 @@ TEST(Expert1CardsGen, EX1_012)
     EXPECT_EQ(curField.GetNumOfMinions(), 0u);
     EXPECT_EQ(curPlayer.currentSpellPower, 0);
     EXPECT_EQ(curPlayer.GetHand().GetNumOfCards(), 6u);
+}
+
+TEST(Expert1CardsGen, CS1_129)
+{
+//  set environment PRIEST vs PALADIN
```
Comment (utilForever) [inline Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Remove comment.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp [hunk 4]
```diff
@@ -198,4 +198,41 @@ TEST(Expert1CardsGen, EX1_012)
...
+    game.StartGame();
+    game.ProcessUntil(Step::MAIN_START);
+
+    Player& curPlayer = game.GetCurrentPlayer();
+    Player& opPlayer = game.GetOpponentPlayer();
+    curPlayer.SetTotalMana(10);
+    curPlayer.SetUsedMana(0);
+    opPlayer.SetTotalMana(10);
+    opPlayer.SetUsedMana(0);
+
```
Comment (utilForever) [inline Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Remove a line.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp [hunk 5]
```diff
@@ -198,4 +198,41 @@ TEST(Expert1CardsGen, EX1_012)
...
+    Player& curPlayer = game.GetCurrentPlayer();
+    Player& opPlayer = game.GetOpponentPlayer();
+    curPlayer.SetTotalMana(10);
+    curPlayer.SetUsedMana(0);
+    opPlayer.SetTotalMana(10);
+    opPlayer.SetUsedMana(0);
+    
+
+    auto& curField = curPlayer.GetField();
```
Comment (utilForever) [inline Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Please put a line down.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp [hunk 6]
```diff
@@ -198,4 +198,41 @@ TEST(Expert1CardsGen, EX1_012)
...
+    const auto card1 = Generic::DrawCard(
+        curPlayer, Cards::GetInstance().FindCardByName("Inner Fire"));
+    const auto card2 = Generic::DrawCard(
+        curPlayer, Cards::GetInstance().FindCardByName("Northshire Cleric"));    
+
+//     1. summon Monion "Northshire Cleric" to cur player field
```
Comment (utilForever) [inline Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Remove comment.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp [hunk 7]
```diff
@@ -198,4 +198,41 @@ TEST(Expert1CardsGen, EX1_012)
...
+    Task::Run(curPlayer, PlayCardTask::Minion(curPlayer, card2));
+//     2. play spell "Inner Fire" to "Northshire Cleric"
```
Comment (utilForever) [inline Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Remove comment.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp [hunk 8]
```diff
@@ -198,4 +198,41 @@ TEST(Expert1CardsGen, EX1_012)
...
+    Task::Run(curPlayer, PlayCardTask::SpellTarget(curPlayer, card1, card2));
+    // assert - "Northshire Cleric" Atk equals Health
```
Comment (utilForever) [inline Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Remove comment.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp [hunk 9]
```diff
@@ -198,4 +198,41 @@ TEST(Expert1CardsGen, EX1_012)
...
+    EXPECT_EQ(curField.GetNumOfMinions(),1);
+    EXPECT_EQ(curField.GetMinion(0)->GetHealth() , curField.GetMinion(0)->GetAttack());
```
Comment (utilForever) [inline Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Check whitespace. Should `GetHealth(), curField.GetMinion(0)`

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp [hunk 10]
```diff
@@ -198,4 +198,41 @@ TEST(Expert1CardsGen, EX1_012)
...
+        curPlayer, Cards::GetInstance().FindCardByName("Inner Fire"));
+    const auto card2 = Generic::DrawCard(
+        curPlayer, Cards::GetInstance().FindCardByName("Northshire Cleric"));    
+
+//     1. summon Monion "Northshire Cleric" to cur player field
+    Task::Run(curPlayer, PlayCardTask::Minion(curPlayer, card2));
+//     2. play spell "Inner Fire" to "Northshire Cleric"
+    Task::Run(curPlayer, PlayCardTask::SpellTarget(curPlayer, card1, card2));
+    // assert - "Northshire Cleric" Atk equals Health
+    EXPECT_EQ(curField.GetNumOfMinions(),1);
```
Comment (utilForever) [inline Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Check whitespace. Should be `EXPECT_EQ(curField.GetNumOfMinions(), 1);`

File: Sources/Rosetta/CardSets/Expert1CardsGen.cpp [hunk 11]
```diff
@@ -95,12 +105,36 @@ void Expert1CardsGen::AddPaladinNonCollect(std::map<std::string, Power>& cards)
...
+    power.AddPowerTask(new GetGameTagTask(EntityType::TARGET, GameTag::HEALTH));
+    power.AddPowerTask(new MathSubTask(EntityType::TARGET, GameTag::DAMAGE));
+    power.AddPowerTask(new AddEnchantmentTask("CS1_129e", EntityType::TARGET));
+    cards.emplace("CS1_129", power);
 }
 
 void Expert1CardsGen::AddPriestNonCollect(std::map<std::string, Power>& cards)
 {
-    (void)cards;
+    Power power;
```
Comment (utilForever) [inline Sources/Rosetta/CardSets/Expert1CardsGen.cpp]: Please put a line down.

File: Sources/Rosetta/CardSets/Expert1CardsGen.cpp [hunk 12]
```diff
@@ -95,12 +105,36 @@ void Expert1CardsGen::AddPaladinNonCollect(std::map<std::string, Power>& cards)
 
 void Expert1CardsGen::AddPriest(std::map<std::string, Power>& cards)
 {
-    (void)cards;
+    Power power;
```
Comment (utilForever) [inline Sources/Rosetta/CardSets/Expert1CardsGen.cpp]: Please put a line down.

**Comment by 95D** [comment]
I will do check today after 7pm!
._.)o

**Comment by utilForever** [comment]
@95D BTW, your time zone is UTC?

**Comment by 95D** [comment]
Yes i am that is UTC/ GMT +9

File: Includes/Rosetta/Enchants/Enchants.hpp (hunk 1 above)
Comment (utilForever) [reply Includes/Rosetta/Enchants/Enchants.hpp]: Fixed.

File: Sources/Rosetta/CardSets/Expert1CardsGen.cpp (hunk 2 above)
Comment (utilForever) [reply Sources/Rosetta/CardSets/Expert1CardsGen.cpp]: Fixed.

File: Sources/Rosetta/CardSets/Expert1CardsGen.cpp (hunk 12 above)
Comment (utilForever) [reply Sources/Rosetta/CardSets/Expert1CardsGen.cpp]: Fixed.

File: Sources/Rosetta/CardSets/Expert1CardsGen.cpp (hunk 11 above)
Comment (utilForever) [reply Sources/Rosetta/CardSets/Expert1CardsGen.cpp]: Fixed.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp (hunk 6 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Fixed.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp (hunk 7 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Fixed.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp (hunk 8 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Fixed.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp (hunk 10 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Fixed.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp (hunk 3 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Fixed.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp (hunk 9 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Fixed and confirmed.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp (hunk 4 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: Fixed.

File: Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp (hunk 5 above)
Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: @95D Please check it. :)

Comment (utilForever) [reply Tests/UnitTests/CardSets/Expert1CardsGenTests.cpp]: I fixed it.

File: Sources/Rosetta/CardSets/Expert1CardsGen.cpp [hunk 13]
```diff
@@ -3,10 +3,19 @@
 // RosettaStone is hearthstone simulator using C++ with reinforcement learning.
 // Copyright (c) 2019 Chris Ohk, Youngjoong Kim, SeungHyun Jeon
 
+#include <Rosetta/Cards/Cards.hpp>
+#include <Rosetta/CardSets/CoreCardsGen.hpp>
```
Comment (CreatorSeraph) [inline Sources/Rosetta/CardSets/Expert1CardsGen.cpp]: Maybe it seems to be unnecessary header.

**Comment by utilForever** [comment]
I fixed code to save your time. :)

File: Sources/Rosetta/CardSets/Expert1CardsGen.cpp (hunk 13 above)
Comment (utilForever) [reply Sources/Rosetta/CardSets/Expert1CardsGen.cpp]: Fixed.

**Comment by 95D** [comment]
Oh Thanks!

**Pull request merged**